File: C/AtomFolder/atom.c
#include <SDL2/SDL.h>
#include <SDL2/SDL_image.h>
#include <SDL2/SDL_ttf.h>
#include <stdio.h> 
#include <stdlib.h>
#include <time.h>

#define WINDOW_WIDTH  2000
#define WINDOW_HEIGHT 600
#define NUM_PARTICLES 2000

typedef struct {
    float x, y;      // Current positio
    float vx, vy;    // Velocity
    int size;
    SDL_Color color;
} Particle;

int main(int argc, char *argv[]) {
    // Seed random number generator
    srand((unsigned int)time(NULL));

    // Initialize SDL
    if (SDL_Init(SDL_INIT_VIDEO) < 0) {
        SDL_Log("Failed to initialize SDL: %s", SDL_GetError());
        return 1;
    }

    // Create an SDL window
    SDL_Window *window = SDL_CreateWindow(
        "Bouncing Particles",
        SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED,
        WINDOW_WIDTH, WINDOW_HEIGHT,
        SDL_WINDOW_SHOWN
    );
    if (!window) {
        SDL_Log("Failed to create window: %s", SDL_GetError());
        SDL_Quit();
        return 1;
    }

    // Create a renderer for the window
    SDL_Renderer *renderer = SDL_CreateRenderer(
        window, -1, SDL_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC
    );
    if (!renderer) {
        SDL_Log("Failed to create renderer: %s", SDL_GetError());
        SDL_DestroyWindow(window);
        SDL_Quit();
        return 1;
    }

    // Create and initialize particles
    Particle particles[NUM_PARTICLES];
    for (int i = 0; i < NUM_PARTICLES; i++) {
        particles[i].x = (float)(rand() % WINDOW_WIDTH);
        particles[i].y = (float)(rand() % WINDOW_HEIGHT);
        // Random velocity between -5 and 5
        particles[i].vx = (float)((rand() % 101) / 10.0f - 5.0f);
        particles[i].vy = (float)((rand() % 101) / 10.0f - 5.0f);
        particles[i].size = rand() % 5 + 5; // Radius between 5 and 9
        // Give each particle a random colour
        particles[i].color.r = rand() % 256;
        particles[i].color.g = rand() % 256;
        particles[i].color.b = rand() % 256;
        particles[i].color.a = 255;
    }

    // Main loop flag
    int running = 1;
    // Frame rate limiter (optional)
    const int frameDelay = 16; // ~60 FPS
    Uint32 frameStart;
    int frameTime;

    while (running) {
        frameStart = SDL_GetTicks();

        // Event handling
        SDL_Event event;
        while (SDL_PollEvent(&event)) {
            if (event.type == SDL_QUIT) {
                running = 0;
            }
            if (event.type == SDL_KEYDOWN) {
                if (event.key.keysym.sym == SDLK_ESCAPE) {
                    running = 0;
                }
            }
        }

        // Update particle positions
        for (int i = 0; i < NUM_PARTICLES; i++) {
            // Move particle
            particles[i].x += particles[i].vx;
            particles[i].y += particles[i].vy;

            // Check collision with window boundaries
            if (particles[i].x < 0) {
                particles[i].x = 0;
                particles[i].vx = -particles[i].vx;
            } else if (particles[i].x > WINDOW_WIDTH - particles[i].size) {
                particles[i].x = WINDOW_WIDTH - particles[i].size;
                particles[i].vx = -particles[i].vx;
            }
            if (particles[i].y < 0) {
                particles[i].y = 0;
                particles[i].vy = -particles[i].vy;
            } else if (particles[i].y > WINDOW_HEIGHT - particles[i].size) {
                particles[i].y = WINDOW_HEIGHT - particles[i].size;
                particles[i].vy = -particles[i].vy;
            }
        }

        // Clear the screen
        SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255);
        SDL_RenderClear(renderer);

        // Render each particle
        for (int i = 0; i < NUM_PARTICLES; i++) {
            SDL_SetRenderDrawColor(renderer,
                                   particles[i].color.r,
                                   particles[i].color.g,
                                   particles[i].color.b,
                                   255);
            // Draw a small filled square for simplicity (or you can draw circles)
            SDL_Rect rect = {
                (int)particles[i].x,
                (int)particles[i].y,
                particles[i].size,
                particles[i].size
            };
            SDL_RenderFillRect(renderer, &rect);
        }

        // Present the updated frame
        SDL_RenderPresent(renderer);

        // Simple frame rate control (optional)
        frameTime = SDL_GetTicks() - frameStart;
        if (frameDelay > frameTime) {
            SDL_Delay(frameDelay - frameTime);
        }
    }

    // Cleanup
    SDL_DestroyRenderer(renderer);
    SDL_DestroyWindow(window);
    SDL_Quit();

    return 0;
}

File: C/AtomFolder/atom.py
import pygame
import random
import sys

# Window dimensions
WINDOW_WIDTH = 1000
WINDOW_HEIGHT = 1000

# Number of particles
NUM_PARTICLES = 50

class Particle:
    def __init__(self):
        # Random initial position
        self.x = random.uniform(0, WINDOW_WIDTH)
        self.y = random.uniform(0, WINDOW_HEIGHT)
        # Random velocity between -5 and 5
        self.vx = random.uniform(-5, 5)
        self.vy = random.uniform(-5, 5)
        # Particle size (radius)
        self.size = random.randint(5, 10)
        # Random colour
        self.color = (
            random.randint(0, 255),
            random.randint(0, 255),
            random.randint(0, 255)
        )

    def update(self):
        # Update position
        self.x += self.vx
        self.y += self.vy
        self.vx = (self.vx -1) * 0.5
        self.vx = (self.vx -1) * 0.5

        # Bounce off left or right edges
        if self.x < 0:
            self.x = 0
            self.vx = -self.vx
            self.vx = (self.vx -1) * 0.5
            self.vx = (self.vx -1) * 0.5
        elif self.x > WINDOW_WIDTH - self.size:
            self.x = WINDOW_WIDTH - self.size
            self.vx = -self.vx
            self.vx = (self.vx -1) * 0.5
            self.vx = (self.vx -1) * 0.5
        # Bounce off top or bottom edges
        if self.y < 0:
            self.y = 0
            self.vy = -self.vy
            self.vx = (self.vx -1) * 0.5
            self.vx = (self.vx -1) * 0.5
        elif self.y > WINDOW_HEIGHT - self.size:
            self.y = WINDOW_HEIGHT - self.size
            self.vy = -self.vy
            self.vx = (self.vx -1) * 0.5
            self.vx = (self.vx -1) * 0.5
    def draw(self, surface):
        # Draw particle as a filled circle
        pygame.draw.circle(surface, self.color, (int(self.x), int(self.y)), self.size)

def main():
    # Initialise Pygame
    pygame.init()

    # Create a window
    screen = pygame.display.set_mode((WINDOW_WIDTH, WINDOW_HEIGHT))
    pygame.display.set_caption("Bouncing Particles")

    # Create a clock to control the frame rate
    clock = pygame.time.Clock()

    # Create a list of particles
    particles = [Particle() for _ in range(NUM_PARTICLES)]

    # Main loop
    running = True
    while running:
        # Limit the frame rate (60 FPS)
        clock.tick(6000)

        # Handle events
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_ESCAPE:
                    running = False

        # Update all particles
        for particle in particles:
            particle.update()

        # Clear screen (fill with black)
        screen.fill((0, 0, 0))

        # Draw all particles
        for particle in particles:
            particle.draw(screen)

        # Flip the display buffer
        pygame.display.flip()

    pygame.quit()
    sys.exit()

if __name__ == "__main__":
    main()
